### ISSUE TITLE
inconsistent bugs

### ISSUE THREAD

**Issue by jerrytesting** [canonical/raft#390]
We found one inconsistent bug in the latest raft version. Here is the bug log: [inconsistent-bugs-canonical-raft](https://github.com/jerrytesting/inconsistent-bugs-canonical-raft) and hope it could help you debug.

**Comment by jerrytesting** [comment]
Besides, we also found one consistency issue about not read committed: [https://github.com/jerrytesting/inconsistent-bugs-canonical-raft/tree/main/not_read_committed](https://github.com/jerrytesting/inconsistent-bugs-canonical-raft/tree/main/not_read_committed)

**Comment by jerrytesting** [comment]
Sorry that our Jepsen setup is not open-sourced now but it will be soon. Currently, I can only give you enough bug reports to help you figure out the issues.

**Comment by jerrytesting** [comment]
Yes, included the assertion checker and reported two assertion failures here: [#291](https://github.com/canonical/raft/issues/291). We are still running to see whether we can capture other issues.

**Comment by MathieuBordere** [comment]
hey @jerrytesting do you have the `app.log` for the nodes for `not_read_uncommitted` like you have for the other bug in the repo?

**Comment by MathieuBordere** [comment]
Also, which version of raft are you using? Based on the log messages,you are not on [v0.17.1](https://github.com/canonical/raft/releases/tag/v0.17.1) which contains quite some bugfixes.

**Comment by MathieuBordere** [comment]
What is also strange is e.g. in `not_read_uncommitted` in `n1`'s app.log, the node starts around log line 88690 as a single node in a cluster and it's accepting requests from Jepsen, timestamp `2023/02/13 14:58:56.631530`.

An example of an inconsistent read is key 55 around timestamp 2023-02-13 14:59:41, that's also a time during which node n1 is accepting requests as a single member and leader of a cluster, but during that time `n3` is the leader of another cluster. So you have effectively 2 clusters running. 

Is it possible you are not setting up the cluster correctly in your tests?

So you effectively have 2 clusters

**Comment by nurturenature** [comment]
Looking at the logs from a Jepsen test perspective, it's hard to make sense of what's happening.
My interpretation is most likely a synthetic, i.e. test environment, failure.

From looking at the log messages, the `dqlite.jepsen` tests are quite old, >= 8 months.

The test framework `mediator` wrappers that were added to the core of Jepsen break nemesis behavior, impact,and intent.

`jepsen.mediator.wrapper.MediatorGen` is not behaving like a generator: 
- some nemesis operations are not `invoke!`d
- operations are emitted duplicated, out of order
- dispatches some `net` operations to `mediator.wrapper` `net` and some straight to Jepsen's `net`
- repeats some single operation generators indefinitely
- doesn't wrap `:final-generator` for healing

Running the current `jepsen.dqlite` with the same test opts did not reproduce the issue.

Note: Jepsen provides extension mechanisms, e.g. `net` protocol, `gen/map`, etc that will do what's visible in the logs without the need for a `mediator`.

**Comment by jerrytesting** [comment]
Yeah, there was one bug in our tool that the stop-partition nemesis was not invoked and so not executed. Thanks for finding this bug in our tool. However, the other points you said are not issues. Our tool only decides how to dispatch nemesis but the real generators are still from Jepsen. 

I think this consistency violation really exists but happens in an extreme environment; for example, the network partition doesn't be healed in long time.
